File: ml/src/ungomoku_ml/rules/tactics.py
"""Vectorized tactical cell detection (immediate wins).

Used by the search to force win-now / block-now cells into the candidate set
regardless of policy priors. Self-play otherwise discovers defense very
slowly: blocking cells sit on the opponent's line, which a rush-biased policy
never proposes, so they rarely enter the sampled root arms.
"""

import numpy as np

from ungomoku_ml.rules.board import EMPTY
from ungomoku_ml.rules.constants import WIN_LENGTH

_DIRECTIONS = ((1, 0), (0, 1), (1, 1), (1, -1))


def _shift(mask: np.ndarray, dx: int, dy: int) -> np.ndarray:
    """Shifts a (15, 15) bool mask by (dx, dy), zero-filling the border.

    out[y, x] = mask[y + dy, x + dx] (False outside the board).
    """
    out = np.zeros_like(mask)
    h, w = mask.shape
    src_y = slice(max(dy, 0), h + min(dy, 0))
    src_x = slice(max(dx, 0), w + min(dx, 0))
    dst_y = slice(max(-dy, 0), h + min(-dy, 0))
    dst_x = slice(max(-dx, 0), w + min(-dx, 0))
    out[dst_y, dst_x] = mask[src_y, src_x]
    return out


def winning_cells_mask(board: np.ndarray, player: int) -> np.ndarray:
    """(15, 15) bool: empty cells where placing ``player``'s stone wins now."""
    own = board == player
    empty = board == EMPTY
    wins = np.zeros_like(empty)
    for dx, dy in _DIRECTIONS:
        # consecutive own stones extending forward / backward from each cell
        fwd = np.zeros(board.shape, dtype=np.int8)
        bwd = np.zeros(board.shape, dtype=np.int8)
        run_f = np.ones(board.shape, dtype=bool)
        run_b = np.ones(board.shape, dtype=bool)
        for step in range(1, WIN_LENGTH):
            run_f &= _shift(own, dx * step, dy * step)
            run_b &= _shift(own, -dx * step, -dy * step)
            fwd += run_f
            bwd += run_b
        wins |= (1 + fwd + bwd) >= WIN_LENGTH
    return wins & empty


def winning_cells(board: np.ndarray, player: int) -> np.ndarray:
    """Flat indices (y * 15 + x) of immediate winning cells for ``player``."""
    return np.flatnonzero(winning_cells_mask(board, player).reshape(-1))


def _five_windows(board: np.ndarray, player: int):
    """Yields (cells, p_count, empty_cells) for every 5-in-a-row window.

    cells are flat indices; iteration is vectorized per direction with the
    per-window detail loop only over qualifying windows.
    """
    own = (board == player).astype(np.int8)
    empty = (board == EMPTY).astype(np.int8)
    size = board.shape[0]
    flat_index = np.arange(size * size).reshape(size, size)
    for dx, dy in _DIRECTIONS:
        # Window start ranges keeping all 5 cells in bounds.
        x0 = slice(0, size - 4 * dx) if dx > 0 else slice(0, size)
        if dy > 0:
            y0 = slice(0, size - 4 * dy)
        elif dy < 0:
            y0 = slice(4, size)
        else:
            y0 = slice(0, size)
        own_sum = np.zeros_like(own[y0, x0], dtype=np.int8)
        empty_sum = np.zeros_like(own_sum)
        for step in range(WIN_LENGTH):
            sy = slice(y0.start + dy * step, (y0.stop or size) + dy * step)
            sx = slice(x0.start + dx * step, (x0.stop or size) + dx * step)
            own_sum = own_sum + own[sy, sx]
            empty_sum = empty_sum + empty[sy, sx]
        starts_y, starts_x = np.nonzero((own_sum == WIN_LENGTH - 2) & (empty_sum == 2))
        for wy, wx in zip(starts_y, starts_x, strict=True):
            y = int(wy) + (y0.start or 0)
            x = int(wx) + (x0.start or 0)
            empties = []
            for step in range(WIN_LENGTH):
                cy, cx = y + dy * step, x + dx * step
                if board[cy, cx] == EMPTY:
                    empties.append(int(flat_index[cy, cx]))
            yield empties


def double_threat_cells(board: np.ndarray, player: int) -> np.ndarray:
    """Empty cells where placing ``player``'s stone yields >= 2 winning cells.

    A cell with two simultaneous winning threats is (probabilistically) almost
    unstoppable next turn — the core winning pattern of the racing meta.
    Windows with exactly 3 own stones + 2 empties contribute a threat pair:
    placing at one empty makes the other a winning cell.
    """
    existing = winning_cells(board, player)
    threats: dict[int, set[int]] = {}
    for pair in _five_windows(board, player):
        if len(pair) != 2:
            continue
        a, b = pair
        threats.setdefault(a, set()).add(b)
        threats.setdefault(b, set()).add(a)

    existing_set = {int(c) for c in existing}
    out = []
    for cell, raw_threats in threats.items():
        # Winning cells after placing at `cell` = new threats from windows
        # through it, united with pre-existing winning cells (minus itself).
        # The move must CREATE at least one new threat: with >= 2 pre-existing
        # winning cells every placement trivially "has two threats", but those
        # positions are handled by the forced win-cell logic.
        new_threats = raw_threats - existing_set
        if not new_threats:
            continue
        winning_after = new_threats | (existing_set - {cell})
        if len(winning_after) >= 2:
            out.append(cell)
    return np.array(sorted(out), dtype=np.int64)

File: apps/web/src/features/game/lib/ai/tactics.ts
/** Flat-board win detection and tactical cell scanning (mirrors ml rules). */

import { BOARD_SIZE, WIN_LENGTH } from "@pkg/shared/constants";
import { CELLS, cellXY, EMPTY } from "./types";

const DIRECTIONS: [number, number][] = [
  [1, 0],
  [0, 1],
  [1, 1],
  [1, -1],
];

/** Win check on a flat board; mirrors packages/core/win-detection.ts. */
export function checkWinAtFlat(
  board: Int8Array,
  cell: number,
  stone: number,
): boolean {
  const { x, y } = cellXY(cell);
  for (const [dx, dy] of DIRECTIONS) {
    let count = 1;
    for (let i = 1; i < WIN_LENGTH; i++) {
      const cx = x + dx * i;
      const cy = y + dy * i;
      if (
        cx < 0 ||
        cx >= BOARD_SIZE ||
        cy < 0 ||
        cy >= BOARD_SIZE ||
        board[cy * BOARD_SIZE + cx] !== stone
      ) {
        break;
      }
      count++;
    }
    for (let i = 1; i < WIN_LENGTH; i++) {
      const cx = x - dx * i;
      const cy = y - dy * i;
      if (
        cx < 0 ||
        cx >= BOARD_SIZE ||
        cy < 0 ||
        cy >= BOARD_SIZE ||
        board[cy * BOARD_SIZE + cx] !== stone
      ) {
        break;
      }
      count++;
    }
    if (count >= WIN_LENGTH) return true;
  }
  return false;
}

export function isBoardFullFlat(board: Int8Array): boolean {
  for (let i = 0; i < CELLS; i++) {
    if (board[i] === EMPTY) return false;
  }
  return true;
}

/** Empty cells where placing `stone` wins immediately. */
export function winningCellsFlat(board: Int8Array, stone: number): number[] {
  const wins: number[] = [];
  for (let i = 0; i < CELLS; i++) {
    if (board[i] !== EMPTY) continue;
    board[i] = stone;
    if (checkWinAtFlat(board, i, stone)) wins.push(i);
    board[i] = EMPTY;
  }
  return wins;
}

const MAX_FORCING_CANDIDATES = 12;

/** Cells that would create at least one NEW winning cell, with their counts. */
function threatCreatingCells(
  board: Int8Array,
  stone: number,
): Map<number, number> {
  const before = new Set(winningCellsFlat(board, stone));
  const out = new Map<number, number>();
  for (let cell = 0; cell < CELLS; cell++) {
    if (board[cell] !== EMPTY) continue;
    board[cell] = stone;
    let created = 0;
    for (const w of winningCellsFlat(board, stone)) {
      if (!before.has(w)) created++;
    }
    board[cell] = EMPTY;
    if (created > 0) out.set(cell, created);
  }
  return out;
}

function forcedWinAfter(
  board: Int8Array,
  stone: number,
  depth: number,
  opponent: number,
): boolean {
  if (depth <= 0) return false;
  if (winningCellsFlat(board, stone).length > 0) return true;
  const threats = threatCreatingCells(board, stone);
  const candidates = [...threats.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, MAX_FORCING_CANDIDATES);
  for (const [cell] of candidates) {
    board[cell] = stone;
    const wins = winningCellsFlat(board, stone);
    let found = false;
    if (wins.length >= 2) {
      found = true;
    } else if (wins.length === 1) {
      const w = wins[0] as number;
      board[w] = opponent;
      if (
        winningCellsFlat(board, opponent).length === 0 &&
        forcedWinAfter(board, stone, depth - 1, opponent)
      ) {
        found = true;
      }
      board[w] = EMPTY;
    }
    board[cell] = EMPTY;
    if (found) return true;
  }
  return false;
}

/**
 * Empty cells initiating a forced win within `depth` forcing moves —
 * win-now cells, double-threat cells (depth 1), and deeper forcing chains.
 * Conservative: a forced block that creates a counter-threat aborts the
 * line. Mirrors ml rules/solver.py forcing_win_cells. Root-only cost.
 */
export function forcingWinCellsFlat(
  board: Int8Array,
  stone: number,
  depth: number,
): number[] {
  if (depth <= 0) return [];
  const opponent = stone === 1 ? 2 : 1;
  const out = new Set<number>(winningCellsFlat(board, stone));
  const threats = threatCreatingCells(board, stone);
  const candidates = [...threats.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, MAX_FORCING_CANDIDATES);
  for (const [cell] of candidates) {
    if (out.has(cell)) continue;
    board[cell] = stone;
    const wins = winningCellsFlat(board, stone);
    if (wins.length >= 2) {
      out.add(cell);
    } else if (wins.length === 1 && depth >= 2) {
      const w = wins[0] as number;
      board[w] = opponent;
      if (
        winningCellsFlat(board, opponent).length === 0 &&
        forcedWinAfter(board, stone, depth - 1, opponent)
      ) {
        out.add(cell);
      }
      board[w] = EMPTY;
    }
    board[cell] = EMPTY;
  }
  return [...out].sort((a, b) => a - b);
}

/**
 * Empty cells where placing `stone` yields two or more simultaneous winning
 * cells (a double threat — almost unstoppable next turn). Root-only cost:
 * one full scan per move. Mirrors ml rules/tactics.py double_threat_cells.
 */
export function doubleThreatCellsFlat(
  board: Int8Array,
  stone: number,
): number[] {
  const before = new Set(winningCellsFlat(board, stone));
  const result: number[] = [];
  for (let cell = 0; cell < CELLS; cell++) {
    if (board[cell] !== EMPTY) continue;
    board[cell] = stone;
    let threats = 0;
    let created = false;
    for (let w = 0; w < CELLS && (threats < 2 || !created); w++) {
      if (board[w] !== EMPTY) continue;
      board[w] = stone;
      if (checkWinAtFlat(board, w, stone)) {
        threats++;
        if (!before.has(w)) created = true;
      }
      board[w] = EMPTY;
    }
    board[cell] = EMPTY;
    // Must create at least one NEW threat and end with two simultaneously
    // (positions with two pre-existing wins are the forced-win logic's job).
    if (threats >= 2 && created) result.push(cell);
  }
  return result;
}
